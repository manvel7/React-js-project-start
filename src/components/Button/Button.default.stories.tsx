import React from "react"
import {Meta, Story} from "@storybook/react/types-6-0"
import Button, {IProps} from "./Button"

const meta: Meta<IProps> = {
  title: "Components/Button/Default",
  component: Button,
  argTypes: {
    onClick: {action: "onClick"}
  }
}

const ButtonStory: Story<IProps> = (args) => <Button {...args}>Button</Button>

const Large = ButtonStory.bind({})
Large.args = {
  size: "lg",
  variant: "default"
}

const LargeDisabled = ButtonStory.bind({})
LargeDisabled.args = {
  size: "lg",
  variant: "default",
  isDisabled: true
}

const Medium = ButtonStory.bind({})
Medium.args = {
  size: "md",
  variant: "default"
}

const MediumDisabled = ButtonStory.bind({})
MediumDisabled.args = {
  size: "md",
  variant: "default",
  isDisabled: true
}

const Small = ButtonStory.bind({})
Small.args = {
  size: "sm",
  variant: "default"
}

const SmallDisabled = ButtonStory.bind({})
SmallDisabled.args = {
  size: "sm",
  variant: "default",
  isDisabled: true
}

const Info = ButtonStory.bind({})
Info.args = {
  state: "info"
}

const Success = ButtonStory.bind({})
Success.args = {
  state: "success"
}

const Warn = ButtonStory.bind({})
Warn.args = {
  state: "warn"
}

const Error = ButtonStory.bind({})
Error.args = {
  state: "error"
}

export {
  Large,
  LargeDisabled,
  Medium,
  MediumDisabled,
  Small,
  SmallDisabled,
  Info,
  Success,
  Warn,
  Error
}

export default meta
