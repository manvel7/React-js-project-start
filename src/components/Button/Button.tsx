import {ButtonHTMLAttributes} from "react"
import styled, {css, StyledProps} from "styled-components"
import {h4Mixin, h6Mixin, smallMixin} from "../../styles/typography"

export type ButtonSize = "sm" | "md" | "lg"
export type ButtonVariant = "default" | "outlined" | "text"
export type ButtonState = "info" | "warn" | "success" | "error"

export interface IProps extends ButtonHTMLAttributes<HTMLButtonElement> {
  isDisabled?: boolean
  size?: ButtonSize
  variant?: ButtonVariant
  state?: ButtonState
}

const lgMixin = css<StyledProps<IProps>>`
  border-radius: 0.8rem;
  padding: 0.8rem 1.6rem;
  ${h4Mixin<IProps>()}
`

const smMixin = css<StyledProps<IProps>>`
  border-radius: 0.3rem;
  padding: 0.2rem 1.6rem;
  ${smallMixin<IProps>()}
`

const mdMixin = css<StyledProps<IProps>>`
  border-radius: 0.4rem;
  padding: 0.6rem 1.6rem;
  ${h6Mixin<IProps>()}
`

const outlinedMixin = css<StyledProps<IProps>>`
  background-color: transparent;
  border: 0.1rem solid
    ${({isDisabled, state, theme}) => {
      if (isDisabled) {
        return theme.palette.disabled
      }

      switch (state) {
        case "error":
          return theme.palette.alert.error
        case "warn":
          return theme.palette.alert.warn
        case "success":
          return theme.palette.alert.success
        case "info":
        default:
          return theme.palette.primary
      }
    }};
  box-sizing: border-box;
  color: ${({isDisabled, theme}) =>
    isDisabled ? theme.palette.text.darkDimmed : theme.palette.primary};
`

const textMixin = css<StyledProps<IProps>>`
  background-color: transparent;
  color: ${({isDisabled, theme}) =>
    isDisabled ? theme.palette.text.darkDimmed : theme.palette.primary};
`

const defaultMixin = css<StyledProps<IProps>>`
  background-color: ${({isDisabled, state, theme}) => {
    if (isDisabled) {
      return theme.palette.background.disabled
    }

    switch (state) {
      case "error":
        return theme.palette.alert.error
      case "warn":
        return theme.palette.alert.warn
      case "success":
        return theme.palette.alert.success
      case "info":
      default:
        return theme.palette.primary
    }
  }};
  color: ${({isDisabled, theme}) =>
    isDisabled ? theme.palette.text.darkDimmed : theme.palette.text.white};
`

const infoMixin = css<StyledProps<IProps>>`
  color: ${({theme}) => theme.palette.alert.info};
`

const warnMixin = css<StyledProps<IProps>>`
  color: ${({theme}) => theme.palette.alert.warn};
`

const errorMixin = css<StyledProps<IProps>>`
  color: ${({theme}) => theme.palette.alert.error};
`

const successMixin = css<StyledProps<IProps>>`
  color: ${({theme}) => theme.palette.alert.success};
`

const Button = styled.button<IProps>`
  border: 0.2rem solid transparent;
  cursor: pointer;

  ${({size}) => {
    switch (size) {
      case "lg":
        return lgMixin
      case "sm":
        return smMixin
      case "md":
      default:
        return mdMixin
    }
  }}

  ${({variant}) => {
    switch (variant) {
      case "outlined":
        return outlinedMixin
      case "text":
        return textMixin
      case "default":
      default:
        return defaultMixin
    }
  }}

  ${({state, variant}) => {
    if (!variant || variant === "default") {
      return ""
    }

    switch (state) {
      case "info":
        return infoMixin
      case "success":
        return successMixin
      case "warn":
        return warnMixin
      case "error":
        return errorMixin
      default:
        return ""
    }
  }}
`

export default Button
