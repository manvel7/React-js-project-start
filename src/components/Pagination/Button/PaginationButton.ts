import {ButtonHTMLAttributes} from "react"
import styled, {css, StyledProps} from "styled-components"
import {h4Mixin, h5Mixin} from "../../../styles/typography"

export type ButtonSize = "md" | "lg"
export type ButtonVariant = "normal" | "circle"

interface IProps extends ButtonHTMLAttributes<HTMLButtonElement> {
  isActive?: boolean
  size?: ButtonSize
  variant?: ButtonVariant
}

const lgMixin = css<StyledProps<IProps>>`
  ${h4Mixin<IProps>()}
`

const mdMixin = css<StyledProps<IProps>>`
  ${h5Mixin<IProps>()}
`

const normalMixin = css<StyledProps<IProps>>`
  border-radius: 0.4rem;
`

const circleMixin = css<StyledProps<IProps>>`
  /*width is not fixed, so 50% doesn't work*/
  border-radius: 10rem;
`

const activeMixin = css<StyledProps<IProps>>`
  background-color: ${({theme}) => theme.palette.primary};
  color: ${({theme}) => theme.palette.text.white};
`

const PaginationButton = styled.button<IProps>`
  background-color: ${({theme}) => theme.palette.background.tinted};
  border: none;
  color: ${({theme}) => theme.palette.primary};
  cursor: pointer;
  margin: 0 0.4rem;
  min-height: 3.2rem;
  min-width: 3.2rem;
  padding: 0.4rem 0.8rem;

  &[disabled] {
    cursor: initial;
  }

  &:first-child {
    margin-left: 0;
  }

  &:last-child {
    margin-right: 0;
  }

  ${({size}) => {
    switch (size) {
      case "lg":
        return lgMixin
      case "md":
      default:
        return mdMixin
    }
  }}

  ${({variant}) => {
    switch (variant) {
      case "circle":
        return circleMixin
      case "normal":
      default:
        return normalMixin
    }
  }}

  ${({isActive}) => (isActive ? activeMixin : "")}
`

PaginationButton.defaultProps = {
  variant: "normal",
  isActive: false
}

export default PaginationButton
