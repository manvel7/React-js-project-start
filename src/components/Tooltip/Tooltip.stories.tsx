import React from "react"
import {Meta, Story} from "@storybook/react/types-6-0"
import Pill from "../Pill"
import Tooltip, {IProps} from "./Tooltip"

const meta: Meta<IProps> = {
  title: "Components/Tooltip",
  component: Tooltip,
  argTypes: {
    placement: {
      control: {
        type: "select",
        options: ["top", "bottom", "left", "right"]
      }
    },
    trigger: {
      control: {
        type: "select",
        options: ["click", "hover", "none"]
      }
    },
    tooltipShown: {
      control: {
        type: "boolean"
      }
    },
    followCursor: {
      control: {
        type: "boolean"
      }
    }
  }
}

const TooltipStory: Story<IProps> = (args) => <Tooltip {...args} />

const TextReference = TooltipStory.bind({})
TextReference.args = {
  children: "Text reference",
  content: "tooltip"
}

const Manual = TooltipStory.bind({})
Manual.args = {
  children: <Pill>Manual state</Pill>,
  content: "tooltip",
  trigger: "none",
  tooltipShown: true
}

const Placement = TooltipStory.bind({})
Placement.args = {
  children: <Pill>Right placement</Pill>,
  content: "tooltip",
  placement: "right"
}

export {TextReference, Manual, Placement}

export default meta
