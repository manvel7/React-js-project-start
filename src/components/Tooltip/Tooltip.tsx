import React, {
  FunctionComponent,
  HTMLAttributes,
  ReactNode,
  Children,
  isValidElement,
  cloneElement,
  createElement
} from "react"
import TooltipTrigger, {TooltipTriggerProps} from "react-popper-tooltip"
import TooltipArrow from "./Arrow"
import TooltipContainer from "./Container"

export interface IProps
  extends HTMLAttributes<HTMLDivElement>,
    Partial<
      Omit<TooltipTriggerProps, "children" | "onVisibilityChange" | "tooltip">
    > {
  content: ReactNode
}

const Tooltip: FunctionComponent<IProps> = (props) => {
  const {
    children,
    content,
    tooltipShown,
    placement,
    trigger,
    delayHide,
    followCursor,
    defaultTooltipShown
  } = props

  if (Children.count(children) > 1) {
    throw new Error("Tooltip children should contain only 1 element")
  }

  let reference: ReactNode =
    (Array.isArray(children) ? children[0] : children) || createElement("span")

  if (typeof reference === "string") {
    reference = createElement("span", {}, reference)
  }

  return (
    <TooltipTrigger
      tooltipShown={tooltipShown}
      trigger={trigger}
      placement={placement}
      delayHide={delayHide}
      followCursor={followCursor}
      defaultTooltipShown={defaultTooltipShown}
      modifiers={[
        {
          name: "offset",
          options: {
            offset: [0, 10]
          }
        }
      ]}
      tooltip={({
        arrowRef,
        tooltipRef,
        getArrowProps,
        getTooltipProps,
        placement
      }) => (
        <TooltipContainer
          {...getTooltipProps({
            ref: tooltipRef
          })}>
          <TooltipArrow
            {...getArrowProps({
              ref: arrowRef,
              "data-placement": placement
            })}
          />
          {content}
        </TooltipContainer>
      )}>
      {({getTriggerProps, triggerRef}) =>
        cloneElement(
          isValidElement(reference) ? reference : createElement("span"),
          {
            ...getTriggerProps({
              ref: triggerRef
            })
          }
        )
      }
    </TooltipTrigger>
  )
}

Tooltip.defaultProps = {
  placement: "bottom",
  delayHide: 200
}

export default Tooltip
