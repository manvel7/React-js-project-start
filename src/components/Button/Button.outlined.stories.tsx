import React from "react"
import {Meta, Story} from "@storybook/react/types-6-0"
import Button, {IProps} from "./Button"

const meta: Meta<IProps> = {
  title: "Components/Button/Outlined",
  component: Button,
  argTypes: {
    onClick: {action: "onClick"}
  }
}

const ButtonStory: Story<IProps> = (args) => <Button {...args}>Button</Button>

const Large = ButtonStory.bind({})
Large.args = {
  size: "lg",
  variant: "outlined"
}

const LargeDisabled = ButtonStory.bind({})
LargeDisabled.args = {
  size: "lg",
  variant: "outlined",
  isDisabled: true
}

const Medium = ButtonStory.bind({})
Medium.args = {
  size: "md",
  variant: "outlined"
}

const MediumDisabled = ButtonStory.bind({})
MediumDisabled.args = {
  size: "md",
  variant: "outlined",
  isDisabled: true
}

const Small = ButtonStory.bind({})
Small.args = {
  size: "sm",
  variant: "outlined"
}

const SmallDisabled = ButtonStory.bind({})
SmallDisabled.args = {
  size: "sm",
  variant: "outlined",
  isDisabled: true
}

const Info = ButtonStory.bind({})
Info.args = {
  variant: "outlined",
  state: "info"
}

const Success = ButtonStory.bind({})
Success.args = {
  variant: "outlined",
  state: "success"
}

const Warn = ButtonStory.bind({})
Warn.args = {
  variant: "outlined",
  state: "warn"
}

const Error = ButtonStory.bind({})
Error.args = {
  variant: "outlined",
  state: "error"
}

export {
  Large,
  LargeDisabled,
  Medium,
  MediumDisabled,
  Small,
  SmallDisabled,
  Info,
  Success,
  Warn,
  Error
}

export default meta
