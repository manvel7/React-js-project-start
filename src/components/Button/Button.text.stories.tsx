import React from "react"
import {Meta, Story} from "@storybook/react/types-6-0"
import Button, {IProps} from "./Button"

const meta: Meta<IProps> = {
  title: "Components/Button/Text",
  component: Button,
  argTypes: {
    onClick: {action: "onClick"}
  }
}

const ButtonStory: Story<IProps> = (args) => <Button {...args}>Button</Button>

const Large = ButtonStory.bind({})
Large.args = {
  size: "lg",
  variant: "text"
}

const LargeDisabled = ButtonStory.bind({})
LargeDisabled.args = {
  size: "lg",
  variant: "text",
  isDisabled: true
}

const Medium = ButtonStory.bind({})
Medium.args = {
  size: "md",
  variant: "text"
}

const MediumDisabled = ButtonStory.bind({})
MediumDisabled.args = {
  size: "md",
  variant: "text",
  isDisabled: true
}

const Small = ButtonStory.bind({})
Small.args = {
  size: "sm",
  variant: "text"
}

const SmallDisabled = ButtonStory.bind({})
SmallDisabled.args = {
  size: "sm",
  variant: "text",
  isDisabled: true
}

const Info = ButtonStory.bind({})
Info.args = {
  variant: "text",
  state: "info"
}

const Success = ButtonStory.bind({})
Success.args = {
  variant: "text",
  state: "success"
}

const Warn = ButtonStory.bind({})
Warn.args = {
  variant: "text",
  state: "warn"
}

const Error = ButtonStory.bind({})
Error.args = {
  variant: "text",
  state: "error"
}

export {
  Large,
  LargeDisabled,
  Medium,
  MediumDisabled,
  Small,
  SmallDisabled,
  Info,
  Success,
  Warn,
  Error
}

export default meta
