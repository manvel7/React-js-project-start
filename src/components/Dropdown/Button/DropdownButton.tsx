import React, {FunctionComponent} from "react"
import {faCaretDown} from "@fortawesome/pro-solid-svg-icons"
import {FontAwesomeIcon} from "@fortawesome/react-fontawesome"
import Button from "../../Button"
import ButtonGroup from "../../ButtonGroup"
import {IProps as IDropdownProps} from "../Dropdown"

interface IProps extends IDropdownProps {
  isOpened: boolean
  onToggleOpen: () => void
}

const DropdownButton: FunctionComponent<IProps> = ({
  size,
  onButtonClick,
  onToggleOpen,
  title,
  isOpened
}) => (
  <ButtonGroup>
    <Button size={size} onClick={onButtonClick}>
      {title}
    </Button>
    <Button size={size} onClick={onToggleOpen}>
      <FontAwesomeIcon
        icon={faCaretDown}
        size="sm"
        rotation={isOpened ? 180 : undefined}
      />
    </Button>
  </ButtonGroup>
)

export default DropdownButton
