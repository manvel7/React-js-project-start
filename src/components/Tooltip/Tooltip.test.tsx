import React from "react"
import {screen, render} from "@testing-library/react"
import {UiProvider} from "../../styles"
import Button from "../Button"
import Tooltip from "./Tooltip"

describe("Tooltip", () => {
  describe("with text reference", () => {
    it("should match snapshot", () => {
      const result = render(
        <UiProvider>
          <Tooltip content="tooltip content">Text reference</Tooltip>
        </UiProvider>
      )
      expect(result.container.firstChild).toMatchSnapshot()
    })

    it("renders on screen properly", () => {
      render(
        <UiProvider>
          <Tooltip trigger="none" tooltipShown={true} content="tooltip content">
            Text reference
          </Tooltip>
        </UiProvider>
      )

      expect(screen.getByText("tooltip content")).toBeInTheDocument()
    })
  })

  describe("with button reference", () => {
    it("should match snapshot", () => {
      const result = render(
        <UiProvider>
          <Tooltip content="tooltip content">
            <Button>Button reference</Button>
          </Tooltip>
        </UiProvider>
      )
      expect(result.container.firstChild).toMatchSnapshot()
    })
  })

  describe("with manual state", () => {
    it("should match snapshot", () => {
      const result = render(
        <UiProvider>
          <Tooltip content="tooltip content" trigger="none" tooltipShown={true}>
            Manual state
          </Tooltip>
        </UiProvider>
      )
      expect(result.container.firstChild).toMatchSnapshot()
    })
  })

  describe("with placement right", () => {
    it("should match snapshot", () => {
      const result = render(
        <UiProvider>
          <Tooltip content="tooltip content" placement="right">
            Placement right
          </Tooltip>
        </UiProvider>
      )
      expect(result.container.firstChild).toMatchSnapshot()
    })
  })
})
