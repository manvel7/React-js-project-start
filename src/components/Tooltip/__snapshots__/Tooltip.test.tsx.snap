// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tooltip with button reference should match snapshot 1`] = `
.c0 {
  border: 0.2rem solid transparent;
  cursor: pointer;
  border-radius: 0.4rem;
  padding: 0.6rem 1.6rem;
  font-size: 1.6rem;
  font-weight: 600;
  -webkit-letter-spacing: 0em;
  -moz-letter-spacing: 0em;
  -ms-letter-spacing: 0em;
  letter-spacing: 0em;
  line-height: 1.6rem;
  background-color: #001254;
  color: #FEFEFE;
}

.c0 b,
.c0 strong {
  font-weight: 800;
}

<button
  class="c0"
>
  Button reference
</button>
`;

exports[`Tooltip with manual state should match snapshot 1`] = `
<span>
  Manual state
</span>
`;

exports[`Tooltip with placement right should match snapshot 1`] = `
<span>
  Placement right
</span>
`;

exports[`Tooltip with text reference should match snapshot 1`] = `
<span>
  Text reference
</span>
`;
